NOnion/Services/TorServiceHost.fs

```diff
@@ -213,70 +238,515 @@ type TorServiceHost
+                    do! dirStream.ConnectToDirectory() |> Async.Ignore
+
+                    let! _response =
+                        TorHttpClient(
+                            dirStream,
+                            Constants.DefaultHttpHost
+                        )
+                            .PostString
+                            (sprintf
+                                "/tor/hs/%i/publish"
+                                Constants.HiddenServices.Version)
+                            (document.ToString())
+
+                    TorLogger.Log(
+                        sprintf
+                            "TorServiceHost: descriptor uploaded to node with identity %s"
+                            directoryToUploadTo
+                    )
 
-                        let random = SecureRandom()
+                    return ()
+                with
+                | :? UnsuccessfulHttpRequestException ->
+                    // During testing, after migration to microdescriptor, we saw instances of
+                    // 404 error msg when trying to publish our descriptors which mean for
+                    // some reason we're trying to upload descriptor to a directory that
+                    // is not a hidden service directory, there is no point in retrying here.
+                    return ()
+                | ex ->
+                    TorLogger.Log(
+                        sprintf
+                            "TorServiceHost: hs descriptor upload failed, ex=%s"
+                            (ex.ToString())
+                    )
 
-                        kpGen.Init(Ed25519KeyGenerationParameters random)
-                        kpGenX.Init(X25519KeyGenerationParameters random)
+                    return!
+                        self.UploadDescriptor
+                            directoryToUploadTo
+                            document
+                            (retry + 1)
+        }
 
-                        kpGenX.GenerateKeyPair(),
-                        kpGen.GenerateKeyPair(),
-                        (kpGen.GenerateKeyPair().Public
-                        :?> Ed25519PublicKeyParameters)
-                            .GetEncoded()
+    member self.BuildAndUploadDescriptor
+        periodNum
+        periodLength
+        srv
+        (networkStatus: NetworkStatusDocument)
+        =
+        async {
+            let blindedPublicKey =
+                HiddenServicesCipher.BuildBlindedPublicKey
+                    (periodNum, periodLength)
+                    (masterPublicKey.GetEncoded())
+
+            let! responsibleDirs =
+                directory.GetResponsibleHiddenServiceDirectories
+                    blindedPublicKey
+                    srv
+                    periodNum
+                    periodLength
+                    Constants.HiddenServices.Hashring.SpreadStore
+
+            let revisionCounter =
+                //PRIVACY: this should be encrypted with an OPE cipher
```

Comment:
PRIVACY is not a searchable keyword, add a FIXME before or after

Reply:
Fixed.